chaincfg/deployment_time_frame.go

```diff
@@ -0,0 +1,181 @@
+package chaincfg
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/btcsuite/btcd/wire"
+)
+
+var (
+	// ErrNoBlockClock is returned when an operation fails due to lack of
+	// synchornization with the current up to date block clock.
+	ErrNoBlockClock = fmt.Errorf("no block clock synchronized")
+)
+
+// BlockClock is an abstraction over the past median time computation. The past
+// median time computation is used in several consensus checks such as CSV, and
+// also BIP 9 version bits. This interface allows callers to abstract away the
+// computation of the past median time from the perspective of a given block
+// header.
+type BlockClock interface {
+	// PastMedianTime returns the past median time from the PoV of the
+	// passed block header. The past median time is the median time of the
+	// 11 blocks prior to the passed block header.
+	PastMedianTime(*wire.BlockHeader) (time.Time, error)
+}
+
+// ConsensusDeploymentStarter determines if a given consensus deployment has
+// started. A deployment has started once according to the current "time", the
+// deployment is eligible for activation once a perquisite condition has
+// passed.
+type ConsensusDeploymentStarter interface {
```

Comment:
Is there ever a deployment that has a different mechanism/implementation for starting and ending? Or could this just be a `ConsensusActivation` with two methods and the extended `ClockConsensusActivation` interface and `MedianTimeActivation` implementation... Would make code in later commits a bit simpler. But non-blocking.

Reply:
> Is there ever a deployment that has a different mechanism/implementation for starting and ending? 

Back when this was made, doing taproot activation with a combination of height/time was an open option. Since then time was used for start/end and height for the min time. The refactors here now aren't _totally_ necessary, but certainly made some of the added unit tests I wrote a bit easier.